server/sonar-server/src/main/java/org/sonar/server/permission/index/PermissionIndexer.java

```diff
@@ -58,153 +56,130 @@
  *   <li>delete project orphans from index</li>
  * </ul>
  */
-public class PermissionIndexer implements Startable {
+public class PermissionIndexer implements ProjectIndexer, Startable {
 
-  private static final int MAX_BATCH_SIZE = 1000;
+  @VisibleForTesting
+  static final int MAX_BATCH_SIZE = 1000;
 
   private static final String BULK_ERROR_MESSAGE = "Fail to index authorization";
 
   private final ThreadPoolExecutor executor;
   private final DbClient dbClient;
   private final EsClient esClient;
+  private final List<AuthorizationScope> authorizationScopes;
+
+  public PermissionIndexer(DbClient dbClient, EsClient esClient, NeedAuthorizationIndexer[] needAuthorizationIndexers) {
+    this(dbClient, esClient, Arrays.stream(needAuthorizationIndexers)
+      .map(NeedAuthorizationIndexer::getAuthorizationScope)
+      .collect(Collectors.toList(needAuthorizationIndexers.length)));
+  }
 
-  public PermissionIndexer(DbClient dbClient, EsClient esClient) {
+  @VisibleForTesting
+  public PermissionIndexer(DbClient dbClient, EsClient esClient, List<AuthorizationScope> authorizationScopes) {
     this.executor = new ThreadPoolExecutor(0, 1, 0L, TimeUnit.SECONDS, new LinkedBlockingQueue<>());
     this.dbClient = dbClient;
     this.esClient = esClient;
+    this.authorizationScopes = authorizationScopes;
   }
 
-  /**
-   * Index issues authorization and project measures authorization indexes only when they are empty
-   */
   public void indexAllIfEmpty() {
-    Future submit = executor.submit(() -> {
-      if (isIndexEmpty(IssueIndexDefinition.INDEX, IssueIndexDefinition.TYPE_AUTHORIZATION) ||
-        isIndexEmpty(ProjectMeasuresIndexDefinition.INDEX_PROJECT_MEASURES, ProjectMeasuresIndexDefinition.TYPE_AUTHORIZATION) ||
-        isIndexEmpty(ComponentIndexDefinition.INDEX_COMPONENTS, ComponentIndexDefinition.TYPE_AUTHORIZATION)) {
-        truncate(IssueIndexDefinition.INDEX, IssueIndexDefinition.TYPE_AUTHORIZATION);
-        truncate(ProjectMeasuresIndexDefinition.INDEX_PROJECT_MEASURES, ProjectMeasuresIndexDefinition.TYPE_AUTHORIZATION);
-        truncate(ComponentIndexDefinition.INDEX_COMPONENTS, ComponentIndexDefinition.TYPE_AUTHORIZATION);
+    boolean isEmpty = false;
+    for (AuthorizationScope scope : authorizationScopes) {
+      isEmpty |= isAuthorizationTypeEmpty(scope.getIndexName());
+    }
+
+    if (isEmpty) {
+      Future submit = executor.submit(() -> {
+
+        authorizationScopes.forEach(scope -> truncateAuthorizationType(scope.getIndexName()));
```

Comment:
Could be rewritten to `authorizationScopes.map(AuthorizationScope::getIndexName).forEach[Ordered](this::truncateAuthorizationType);`.